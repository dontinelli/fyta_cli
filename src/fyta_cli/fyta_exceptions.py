"""Exceptions for Python client for FYTA."""


class FytaError(Exception):
    """Generic exception."""

class FytaConnectionError(FytaError):
    """Fyta connection exception."""

class FytaAuthentificationError(FytaError):
    """Fyta Password exception (wrong password)."""

class FytaPasswordError(FytaError):
    """Fyta Password exception (wrong password)."""

class FytaPlantError(FytaError):
    """Fyta exception in getting plants."""
