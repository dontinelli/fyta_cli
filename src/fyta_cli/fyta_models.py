"""Models for FYTA."""
from dataclasses import dataclass, field
from datetime import datetime
from enum import IntEnum
from typing import Any

from mashumaro import DataClassDictMixin, field_options


@dataclass
class Credentials():
    """Fyta login model."""
    access_token: str
    expiration: datetime


class PlantStatus(IntEnum):
    """Plant status enum."""
    DELETED = 0
    DOING_GREAT = 1
    NEED_ATTENTION = 2
    NO_SENSOR = 3


class PlantMeasurementStatus(IntEnum):
    """Plant measurement status enum."""
    NO_DATA = 0
    TOO_LOW = 1
    LOW = 2
    PERFECT = 3
    HIGH = 4
    TOO_HIGH = 5


class SensorStatus(IntEnum):
    """Sensor status enum."""
    NONE = 0
    CORRECT = 1
    ERROR = 2


@dataclass
class Plant(DataClassDictMixin):
    """Plant model."""

    # pylint: disable=too-many-instance-attributes

    battery_level: float | None
    fertilise_last: datetime | None
    fertilise_next: datetime | None
    last_updated: datetime | None
    light: float | None
    light_min_acceptable: float | None
    light_min_good: float | None
    light_max_acceptable: float | None
    light_max_good: float | None
    light_status: PlantMeasurementStatus
    low_battery: bool
    name: str = field(metadata=field_options(alias="nickname"))
    moisture: float | None
    moisture_min_acceptable: float | None
    moisture_min_good: float | None
    moisture_max_acceptable: float | None
    moisture_max_good: float | None
    moisture_status: PlantMeasurementStatus
    notification_light: bool
    notification_nutrition: bool
    notification_temperature: bool
    notification_water: bool
    nutrients_status: PlantMeasurementStatus
    sensor_available: bool
    sensor_id: str | None
    sensor_status: SensorStatus
    sensor_update_available: bool
    sw_version: str
    status: PlantStatus
    online: bool
    ph: float | None
    plant_id: int
    plant_origin_path: str
    plant_thumb_path: str
    productive_plant: bool = field(
        metadata=field_options(alias="is_productive_plant"))
    repotted: bool
    salinity: float | None
    salinity_min_acceptable: float | None
    salinity_min_good: float | None
    salinity_max_acceptable: float | None
    salinity_max_good: float | None
    salinity_status: PlantMeasurementStatus
    scientific_name: str
    temperature: float | None
    temperature_min_acceptable: float | None
    temperature_min_good: float | None
    temperature_max_acceptable: float | None
    temperature_max_good: float | None
    temperature_status: PlantMeasurementStatus
    user_picture_path: str = field(metadata=field_options(alias="origin_path"))
    user_thumb_path: str = field(metadata=field_options(alias="thumb_path"))

    @classmethod
    def __pre_deserialize__(cls, d: dict[Any, Any]) -> dict[Any, Any]:

        d |= {"sensor_available": True}
        d |= {"online": True}

        if d.get("measurements") is not None:
            d |= {"battery_level": d["measurements"]["battery"]}
            d |= {"light": d["measurements"]["light"]["values"]["current"]}
            d |= {"light_max_acceptable": d["measurements"]
                  ["light"]["values"]["max_acceptable"]}
            d |= {"light_max_good": d["measurements"]
                  ["light"]["values"]["max_good"]}
            d |= {"light_min_acceptable": d["measurements"]
                  ["light"]["values"]["min_acceptable"]}
            d |= {"light_min_good": d["measurements"]
                  ["light"]["values"]["min_good"]}
            d |= {"light_status": int(
                d["measurements"]["light"].get("status") or 0)}
            d |= {"moisture": d["measurements"]
                  ["moisture"]["values"]["current"]}
            d |= {"moisture_max_acceptable": d["measurements"]
                  ["moisture"]["values"]["max_acceptable"]}
            d |= {"moisture_max_good": d["measurements"]
                  ["moisture"]["values"]["max_good"]}
            d |= {"moisture_min_acceptable": d["measurements"]
                  ["moisture"]["values"]["min_acceptable"]}
            d |= {"moisture_min_good": d["measurements"]
                  ["moisture"]["values"]["min_good"]}
            d |= {"moisture_status": int(
                d["measurements"]["moisture"].get("status") or 0)}
            d |= {"nutrients_status": int(
                d["measurements"]["nutrients"].get("status") or 0)}
            d |= {"ph": d["measurements"].get(
                "ph").get("values").get("current")}
            d |= {"salinity": d["measurements"]
                  ["salinity"]["values"]["current"]}
            d |= {"salinity_max_acceptable": d["measurements"]
                  ["salinity"]["values"]["max_acceptable"]}
            d |= {"salinity_max_good": d["measurements"]
                  ["salinity"]["values"]["max_good"]}
            d |= {"salinity_min_acceptable": d["measurements"]
                  ["salinity"]["values"]["min_acceptable"]}
            d |= {"salinity_min_good": d["measurements"]
                  ["salinity"]["values"]["min_good"]}
            d |= {"salinity_status": int(
                d["measurements"]["salinity"].get("status") or 0)}
            d |= {"temperature": d["measurements"]
                  ["temperature"]["values"]["current"]}
            d |= {"temperature_max_acceptable": d["measurements"]
                  ["temperature"]["values"]["max_acceptable"]}
            d |= {"temperature_max_good": d["measurements"]
                  ["temperature"]["values"]["max_good"]}
            d |= {"temperature_min_acceptable": d["measurements"]
                  ["temperature"]["values"]["min_acceptable"]}
            d |= {"temperature_min_good": d["measurements"]
                  ["temperature"]["values"]["min_good"]}
            d |= {"temperature_status": int(
                d["measurements"]["temperature"].get("status") or 0)}

        if d.get("sensor") is not None:
            d |= {"last_updated": d["sensor"]["received_data_at"]}
            d |= {"low_battery": d["sensor"]["is_battery_low"]}
            d |= {"sensor_id": d["sensor"]["id"]}
            d |= {"sw_version": d["sensor"]["version"]}
        d |= {"sensor_status": int(d.get("sensor", {}).get("status") or 0)}

        d |= {"notification_light": d.get(
            "notifications", {}).get("light", False)}
        d |= {"notification_nutrition": d.get(
            "notifications", {}).get("nutrition", False)}
        d |= {"notification_temperature": d.get(
            "notifications", {}).get("temperature", False)}
        d |= {"notification_water": d.get(
            "notifications", {}).get("water", False)}

        d |= {"fertilise_last": d.get(
            "fertilisation", {}).get("last_fertilised_at")}
        d |= {"fertilise_next": d.get("fertilisation", {}).get("fertilise_at")}
        d |= {"repotted": d.get("fertilisation", {}).get(
            "was_repotted", False)}

        return d
