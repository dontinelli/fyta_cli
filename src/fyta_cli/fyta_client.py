"""Client to access FYTA API."""
# The documentation for the API can be found here:
# https://fyta-io.notion.site/FYTA-Public-API-d2f4c30306f74504924c9a40402a3afd

from __future__ import annotations

from datetime import datetime, timedelta, tzinfo
import logging
from typing import Any

from aiohttp import BasicAuth, ClientSession, ClientTimeout,ClientResponseError

from .fyta_exceptions import (
    FytaConnectionError,
    FytaAuthentificationError,
    FytaPasswordError,
    FytaPlantError,
)
from .fyta_models import Credentials

FYTA_AUTH_URL = "https://web.fyta.de/api/auth/login"
FYTA_PLANT_URL = "https://web.fyta.de/api/user-plant"

_LOGGER = logging.getLogger(__name__)


class Client:
    """Client class to access FYTA API."""

    # pylint: disable=too-many-instance-attributes
    # pylint: disable=too-many-arguments
    # pylint: disable=too-many-positional-arguments

    def __init__(
        self,
        email: str,
        password: str,
        access_token: str,
        expiration: datetime,
        tz: tzinfo,
        session: ClientSession | None,
    ):
        """Initialization."""

        self.email = email
        self.password = password
        self.access_token = access_token
        self.expiration = expiration
        self.timezone = tz

        if not session:
            self.session = ClientSession()
        else:
            self.session = session

        self._close_session = True

        self.request_timeout = 60

    async def test_connection(self) -> bool:
        """Test the connection to FYTA-Server"""

        response = await self.session.post(FYTA_AUTH_URL)
        r = await response.text()

        if r == '{"message": "Unsupported Media Type"}':
            return True

        return False

    async def login(self) -> Credentials:
        """Handle a request to FYTA."""

        if (
            self.access_token != ""
            and self.expiration.timestamp() > datetime.now().timestamp()
        ):
            return Credentials(access_token=self.access_token, expiration=self.expiration)

        payload = {
            "email": self.email,
            "password": self.password,
        }

        try:
            response = await self.session.post(
                url=FYTA_AUTH_URL,
                auth=BasicAuth(self.email, self.password),
                json=payload,
                timeout=ClientTimeout(total=self.request_timeout),
            )

        except TimeoutError as exception:
            _LOGGER.exception("timeout error")
            msg = "Timeout occurred while connecting to Fyta-server"
            raise FytaConnectionError(msg) from exception

        json_response = await response.json()

        if json_response == {"statusCode": 404, "error": "Not Found"}:
            _LOGGER.exception("Authentication failed")
            raise FytaAuthentificationError
        if json_response == {
            "statusCode": 401,
            "error": "Unauthorized",
            "errors": [{"message": "Could not authenticate user"}],
        }:
            raise FytaPasswordError

        self.access_token = json_response["access_token"]
        # self.refresh_token = json_response["refresh_token"]
        self.expiration = datetime.now(self.timezone) + timedelta(
            seconds=int(json_response["expires_in"])
        )

        return Credentials(access_token=self.access_token, expiration=self.expiration)

    async def get_plants(self) -> dict[int, str]:
        """Get a list of all available plants from FYTA"""

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        if self.expiration.timestamp() > datetime.now().timestamp():
            await self.login()  # get new access token, if current token expired

        header = {
            "Authorization": f"Bearer {self.access_token}",
            "Content-Type": "application/json",
        }

        _LOGGER.debug("Try getting list of plants")

        try:
            response = await self.session.get(
                url=FYTA_PLANT_URL,
                headers=header,
                timeout=ClientTimeout(total=self.request_timeout),
            )
        except TimeoutError as exception:
            msg = "Timeout occurred while connecting to Fyta-server"
            raise FytaConnectionError(msg) from exception

        content_type = response.headers.get("Content-Type", "")

        if content_type.count("text/html") > 0:
            text = await response.text()
            msg = "Error occurred while fetching plant data"
            raise FytaPlantError(
                msg,
                {"Content-Type": content_type, "response": text},
            )

        json_response = await response.json()

        plant_list: dict = json_response["plants"]
        _LOGGER.debug("List of plants: %s", plant_list)

        plants: dict[int, str] = {}
        for plant in plant_list:
            plants |= {int(plant["id"]): plant["nickname"]}

        return plants

    async def get_plant_data(self, plant_id: int) -> dict[str, Any]:
        """Get information about a specific plant"""

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        if self.expiration.timestamp() > datetime.now().timestamp():
            await self.login()  # get new access token, if current token expired

        header = {
            "Authorization": f"Bearer {self.access_token}",
            "Content-Type": "application/json",
        }

        url = f"{FYTA_PLANT_URL}/{plant_id}"

        _LOGGER.debug("Try getting data for plant: %s", plant_id)

        try:
            response = await self.session.get(
                url=url,
                headers=header,
                timeout=ClientTimeout(total=self.request_timeout),
            )
        except TimeoutError as exception:
            msg = "Timeout occurred while connecting to Fyta-server"
            raise FytaConnectionError(msg) from exception

        content_type = response.headers.get("Content-Type", "")

        if content_type.count("text/html") > 0:
            text = await response.text()
            msg = f"Error occurred while fetching plant data for plant {plant_id}"
            raise FytaPlantError(
                msg,
                {"Content-Type": content_type, "response": text},
            )

        plant = await response.json()
        _LOGGER.debug("Plant data received: %s", plant)

        return plant

    async def get_plant_image(self, image_url) -> tuple[str | None, bytes] | None:
        """Fetch the user image from the API."""

        if self.expiration.timestamp() > datetime.now().timestamp():
            await self.login()  # get new access token, if current token expired

        header = {
            "Authorization": f"Bearer {self.access_token}",
            "Content-Type": "application/json",
        }

        _LOGGER.debug("Try downloading plant image")

        try:
            response = await self.session.get(
                url=image_url,
                headers=header,
                timeout=ClientTimeout(total=self.request_timeout),
            )
            response.raise_for_status()
        except (TimeoutError, ClientResponseError) as err:
            _LOGGER.debug("Error downloading image: %s", err)
            return None

        _LOGGER.debug("Download of image successful")

        content_type=response.headers.get("Content-Type")

        if response.status != 200 or content_type is None:
            _LOGGER.debug("Error downloading image: %s", content_type)
            return None

        return content_type, await response.read()

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()
