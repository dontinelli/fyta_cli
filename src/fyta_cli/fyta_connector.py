"""Connector class to manage access to FYTA API."""

from datetime import datetime, tzinfo, UTC
from zoneinfo import ZoneInfo

from aiohttp import ClientSession

from .fyta_client import Client
from .fyta_models import Credentials, Plant


class FytaConnector:
    """Connector class to access FYTA API."""

    # pylint: disable=too-many-instance-attributes
    # pylint: disable=too-many-arguments
    # pylint: disable=too-many-positional-arguments

    def __init__(
        self,
        email: str,
        password: str,
        access_token: str = "",
        expiration: datetime | None = None,
        tz: str = "",
        session: ClientSession | None = None,
    ) -> None:
        """Initialize connector class."""

        timezone: tzinfo = UTC if tz == "" else ZoneInfo(tz)
        ex: datetime = (
            datetime.now(timezone)
            if expiration is None
            else expiration.astimezone(timezone)
        )
        self.online: bool = False
        self.plant_list: dict[int, str] = {}
        self.plants: dict[int, Plant] = {}

        self.client = Client(email, password, access_token, ex, timezone, session)

    async def test_connection(self) -> bool:
        """Test if connection to FYTA API works."""

        return await self.client.test_connection()

    async def login(self) -> Credentials:
        """Login with credentials to get access token."""

        login = await self.client.login()

        if login != {}:
            self.online = True

        return login

    async def update_plant_list(self) -> dict[int, str]:
        """Get list of all available plants."""

        self.plant_list = await self.client.get_plants()

        return self.plant_list

    async def update_all_plants(self) -> dict[int, Plant]:
        """Get data of all available plants."""

        plants: dict[int, Plant] = {}

        plant_list: dict[int, str] = await self.update_plant_list()

        for plant in plant_list:
            current_plant: Plant | None = await self.update_plant_data(plant)
            if current_plant is not None:
                plants |= {plant: current_plant}

        self.plants = plants

        return plants

    async def update_plant_data(self, plant_id: int) -> Plant | None:
        """Get data of specific plant."""

        p: dict = await self.client.get_plant_data(plant_id)

        if ("plant" not in p) or (p["plant"]["sensor"] is None):
            return None

        plant_data: dict = p["plant"]

        current_plant = Plant.from_dict(plant_data)
        if current_plant.last_updated is not None:
            current_plant.last_updated = current_plant.last_updated.astimezone(self.client.timezone)

        return current_plant

    async def get_plant_image(self, image_url) -> tuple[str | None, bytes] | None:
        """Fetch the user image from the API."""
        return await self.client.get_plant_image(image_url)

    @property
    def access_token(self) -> str:
        """Access token for FYTA API."""
        return self.client.access_token

    @property
    def data(self) -> dict[int, Plant]:
        """ID for FYTA object."""
        return self.plants

    @property
    def email(self) -> str:
        """Email of FYTA account."""
        return self.client.email

    @property
    def expiration(self) -> datetime:
        """Expiration of access token."""
        return self.client.expiration

    @property
    def fyta_id(self) -> str:
        """ID for FYTA object."""
        return self.email
