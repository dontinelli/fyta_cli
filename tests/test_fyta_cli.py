"""Tests for fyta_cli."""

from datetime import datetime, timedelta, UTC

from aioresponses import aioresponses

import pytest
from syrupy.assertion import SnapshotAssertion

from fyta_cli.fyta_client import FYTA_AUTH_URL, FYTA_PLANT_URL
from fyta_cli.fyta_connector import FytaConnector
from fyta_cli.fyta_exceptions import (
    FytaAuthentificationError,
    FytaConnectionError,
    FytaError,
    FytaPasswordError,
    FytaPlantError,
)
from fyta_cli.fyta_models import Credentials, Plant

from . import load_fixture


@pytest.mark.parametrize(
    ("response", "return_value"),
    [
        ('{"message": "Unsupported Media Type"}', True),
        ('', False),
    ],
)
async def test_connection(
    responses: aioresponses,
    response: str,
    return_value: bool,
) -> None:
    """Test connection."""
    responses.post(
        FYTA_AUTH_URL,
        body=response,
    )

    fyta_connector = FytaConnector("example@example.com", "examplepassword")

    assert await fyta_connector.test_connection() == return_value

    assert fyta_connector.client.session is not None
    assert not fyta_connector.client.session.closed
    await fyta_connector.client.close()
    assert fyta_connector.client.session.closed


async def test_login(
    responses: aioresponses,
) -> None:
    """Test login."""
    responses.post(
        FYTA_AUTH_URL,
        status=200,
        body=load_fixture("login_response.json"),
    )

    fyta_connector = FytaConnector("example@example.com", "examplepassword")

    credentials: Credentials = await fyta_connector.login()

    assert credentials.access_token == "111111111111111111111111111111111111111"
    assert fyta_connector.email == "example@example.com"
    assert fyta_connector.fyta_id == "example@example.com"

    assert fyta_connector.client.session is not None
    assert not fyta_connector.client.session.closed
    await fyta_connector.client.close()
    assert fyta_connector.client.session.closed


async def test_login_existing_credentials(
) -> None:
    """Test login."""

    current_expiration: datetime = datetime.now() + timedelta(days=1)

    fyta_connector = FytaConnector(
        "example@example.com",
        "examplepassword",
        "111111111111111111111111111111111111111",
        current_expiration
    )

    await fyta_connector.login()

    assert fyta_connector.access_token == "111111111111111111111111111111111111111"
    assert fyta_connector.expiration == current_expiration.astimezone(UTC)

    assert fyta_connector.client.session is not None
    assert not fyta_connector.client.session.closed
    await fyta_connector.client.close()
    assert fyta_connector.client.session.closed


@pytest.mark.parametrize(
    ("response", "request_timeout", "error"),
    [
        ('{"message": "Unsupported Media Type"}', True, FytaConnectionError),
        ('{"statusCode": 404, "error": "Not Found"}', False, FytaAuthentificationError),
        ('{"statusCode": 401,"error": "Unauthorized","errors": [{"message": "Could not authenticate user"}]}', False, FytaPasswordError),  # pylint: disable=line-too-long
    ],
)
async def test_login_exceptions(
    responses: aioresponses,
    response: str,
    request_timeout: bool,
    error: FytaError,
) -> None:
    """Test login."""
    responses.post(
        FYTA_AUTH_URL,
        timeout=request_timeout,
        body=response,
    )

    fyta_connector = FytaConnector("example@example.com", "examplepassword")

    with pytest.raises(error):  # type: ignore [call-overload]
        await fyta_connector.login()

    await fyta_connector.client.close()
    assert fyta_connector.client.session.closed


async def test_get_plant_list(
    responses: aioresponses,
    snapshot: SnapshotAssertion
) -> None:
    """Test login."""
    responses.get(
        FYTA_PLANT_URL,
        status=200,
        body=load_fixture("get_user_plants.json"),
    )

    fyta_connector = FytaConnector(
        "example@example.com",
        "examplepassword",
        "111111111111111111111111111111111111111",
        datetime.now() + timedelta(days=1),
    )

    await fyta_connector.update_plant_list()

    assert fyta_connector.data == snapshot

    await fyta_connector.client.close()
    assert fyta_connector.client.session.closed


@pytest.mark.parametrize(
    ("headers", "request_timeout", "error"),
    [
        ({"Content-Type": ""}, True, FytaConnectionError),
        ({"Content-Type": "text/html"}, False, FytaPlantError),
    ],
)
async def test_get_plant_list_exceptions(
    responses: aioresponses,
    headers: dict[str, str],
    request_timeout: bool,
    error: FytaError,
) -> None:
    """Test login."""
    responses.get(
        FYTA_PLANT_URL,
        timeout=request_timeout,
        headers=headers,
    )

    fyta_connector = FytaConnector(
        "example@example.com",
        "examplepassword",
        "111111111111111111111111111111111111111",
        datetime.now() + timedelta(days=1),
    )
    await fyta_connector.client.close()
    fyta_connector.client.session = None  # type: ignore [assignment]

    with pytest.raises(error):  # type: ignore [call-overload]
        await fyta_connector.update_plant_list()


async def test_get_plant_data(
    responses: aioresponses,
    snapshot: SnapshotAssertion
) -> None:
    """Test login."""
    responses.get(
        FYTA_PLANT_URL,
        status=200,
        body=load_fixture("get_user_plants.json"),
    )
    responses.get(
        FYTA_PLANT_URL + f"/{0}",
        status=200,
        body=load_fixture("get_plant_details_0.json"),
    )
    responses.get(
        FYTA_PLANT_URL + f"/{1}",
        status=200,
        body=load_fixture("get_plant_details_1.json"),
    )
    responses.get(
        FYTA_PLANT_URL + f"/{2}",
        status=200,
        body=load_fixture("get_plant_details_2.json"),
    )
    responses.post(
        FYTA_PLANT_URL + f"/measurements/{0}",
        status=200,
        body=load_fixture("get_measurements.json"),
    )
    responses.post(
        FYTA_PLANT_URL + f"/measurements/{1}",
        status=200,
        body=load_fixture("get_measurements.json"),
    )
    responses.post(
        FYTA_PLANT_URL + f"/measurements/{2}",
        status=200,
        body=load_fixture("get_measurements.json"),
    )

    fyta_connector = FytaConnector(
        "example@example.com",
        "examplepassword",
        "111111111111111111111111111111111111111",
        datetime.now() + timedelta(days=1)
    )
    plants: dict[int, Plant] = await fyta_connector.update_all_plants()

    assert plants == snapshot

    await fyta_connector.client.close()
    assert fyta_connector.client.session.closed


@pytest.mark.parametrize(
    ("headers", "request_timeout", "error"),
    [
        ({"Content-Type": "application/json"}, True, FytaConnectionError),
        ({"Content-Type": "text/html"}, False, FytaPlantError),
    ],
)
async def test_get_plant_data_exceptions(
    responses: aioresponses,
    headers: dict[str, str],
    request_timeout: bool,
    error: FytaError,
) -> None:
    """Test login."""
    responses.get(
        FYTA_PLANT_URL + "/0",
        headers=headers,
        timeout=request_timeout,
        body=load_fixture("get_plant_details_0.json"),
    )
    responses.get(
        FYTA_PLANT_URL + "/1",
        headers=headers,
        timeout=request_timeout,
        body=load_fixture("get_plant_details_1.json"),
    )

    fyta_connector = FytaConnector(
        "example@example.com",
        "examplepassword",
        "111111111111111111111111111111111111111",
        datetime.now() + timedelta(days=1)
    )
    await fyta_connector.client.close()
    fyta_connector.client.session = None  # type: ignore [assignment]

    with pytest.raises(error):  # type: ignore [call-overload]
        await fyta_connector.client.get_plant_data(0)

    with pytest.raises(error):  # type: ignore [call-overload]
        await fyta_connector.client.get_plant_data(1)

    await fyta_connector.client.close()
    assert fyta_connector.client.session.closed

async def test_get_plant_image(
    responses: aioresponses,
) -> None:
    """Test getting plant image."""
    responses.post(
        FYTA_AUTH_URL,
        status=200,
        body=load_fixture("login_response.json"),
    )
    responses.get(
        "https://api.prod.fyta-app.de/user-plant/1/origin_path",
        headers={"Content-Type": "image/png"},
        status=200,
        body=bytes([100]),
    )

    fyta_connector = FytaConnector("example@example.com", "examplepassword")

    response = await fyta_connector.get_plant_image(
        "https://api.prod.fyta-app.de/user-plant/1/origin_path"
    )

    assert response is not None
    (content_type, raw_image) = response

    assert content_type == "image/png"
    assert raw_image == bytes([100])

    await fyta_connector.client.close()
    assert fyta_connector.client.session.closed


async def test_get_plant_image_errors(
    responses: aioresponses,
) -> None:
    """Test update errors."""
    responses.post(
        FYTA_AUTH_URL,
        status=200,
        body=load_fixture("login_response.json"),
    )
    responses.get(
        "https://api.prod.fyta-app.de/user-plant/1/origin_path",
        headers={"Content-Type": "text/html"},
        status=500,
    )
    responses.get(
        "https://api.prod.fyta-app.de/user-plant/1/origin_path",
        headers={"Content-Type": "text/html"},
        status=204,
    )

    fyta_connector = FytaConnector("example@example.com", "examplepassword")

    response = await fyta_connector.get_plant_image(
        "https://api.prod.fyta-app.de/user-plant/1/origin_path"
    )
    assert response is None

    response = await fyta_connector.get_plant_image(
        "https://api.prod.fyta-app.de/user-plant/1/origin_path"
    )
    assert response is None

    await fyta_connector.client.close()
    assert fyta_connector.client.session.closed
