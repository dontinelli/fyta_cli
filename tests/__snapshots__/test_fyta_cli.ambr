# serializer version: 1
# name: test_get_plant_data[asyncio]
  dict({
    0: dict({
      'battery_level': 100.0,
      'fertilise_last': datetime.datetime(2024, 11, 16, 0, 0),
      'fertilise_next': datetime.datetime(2025, 1, 11, 0, 0),
      'last_updated': datetime.datetime(2023, 1, 1, 10, 10, tzinfo=datetime.timezone.utc),
      'light': 2.0,
      'light_max_acceptable': 675.0,
      'light_max_good': 450.0,
      'light_min_acceptable': 18.0,
      'light_min_good': 20.0,
      'light_status': <PlantMeasurementStatus.TOO_LOW: 1>,
      'low_battery': False,
      'moisture': 61.0,
      'moisture_max_acceptable': 80.0,
      'moisture_max_good': 70.0,
      'moisture_min_acceptable': 25.0,
      'moisture_min_good': 35.0,
      'moisture_status': <PlantMeasurementStatus.PERFECT: 3>,
      'name': 'Gummibaum',
      'notification_light': False,
      'notification_nutrition': True,
      'notification_temperature': False,
      'notification_water': True,
      'nutrients_status': <PlantMeasurementStatus.PERFECT: 3>,
      'online': True,
      'ph': None,
      'plant_id': 201,
      'plant_origin_path': '<url>',
      'plant_thumb_path': '<url>',
      'productive_plant': False,
      'repotted': True,
      'salinity': 1.0,
      'salinity_max_acceptable': 1.2,
      'salinity_max_good': 1.0,
      'salinity_min_acceptable': 0.4,
      'salinity_min_good': 0.6,
      'salinity_status': <PlantMeasurementStatus.LOW: 2>,
      'scientific_name': 'Ficus elastica',
      'sensor_available': True,
      'sensor_id': 'AA:AA:AA:2B:AF:F4',
      'sensor_status': <SensorStatus.CORRECT: 1>,
      'sensor_update_available': False,
      'status': <PlantStatus.NEED_ATTENTION: 2>,
      'sw_version': '0.30.0',
      'temperature': 18.0,
      'temperature_max_acceptable': 42.0,
      'temperature_max_good': 36.0,
      'temperature_min_acceptable': 10.0,
      'temperature_min_good': 17.0,
      'temperature_status': <PlantMeasurementStatus.LOW: 2>,
      'user_picture_path': '<url>',
      'user_thumb_path': '<url>',
    }),
    1: dict({
      'battery_level': 50.0,
      'fertilise_last': None,
      'fertilise_next': None,
      'last_updated': datetime.datetime(2023, 1, 1, 10, 10, tzinfo=datetime.timezone.utc),
      'light': None,
      'light_max_acceptable': 675.0,
      'light_max_good': 450.0,
      'light_min_acceptable': 18.0,
      'light_min_good': 20.0,
      'light_status': <PlantMeasurementStatus.NO_DATA: 0>,
      'low_battery': False,
      'moisture': 61.0,
      'moisture_max_acceptable': 80.0,
      'moisture_max_good': 70.0,
      'moisture_min_acceptable': 25.0,
      'moisture_min_good': 35.0,
      'moisture_status': <PlantMeasurementStatus.PERFECT: 3>,
      'name': 'Kakaobaum',
      'notification_light': False,
      'notification_nutrition': False,
      'notification_temperature': False,
      'notification_water': False,
      'nutrients_status': <PlantMeasurementStatus.PERFECT: 3>,
      'online': True,
      'ph': None,
      'plant_id': 201,
      'plant_origin_path': '<url>',
      'plant_thumb_path': '<url>',
      'productive_plant': True,
      'repotted': False,
      'salinity': 1.0,
      'salinity_max_acceptable': 1.2,
      'salinity_max_good': 1.0,
      'salinity_min_acceptable': 0.4,
      'salinity_min_good': 0.6,
      'salinity_status': <PlantMeasurementStatus.LOW: 2>,
      'scientific_name': 'Theobroma cacao',
      'sensor_available': True,
      'sensor_id': 'AA:AA:AA:2B:AF:F4',
      'sensor_status': <SensorStatus.CORRECT: 1>,
      'sensor_update_available': False,
      'status': <PlantStatus.NEED_ATTENTION: 2>,
      'sw_version': '0.30.0',
      'temperature': None,
      'temperature_max_acceptable': 42.0,
      'temperature_max_good': 36.0,
      'temperature_min_acceptable': 10.0,
      'temperature_min_good': 17.0,
      'temperature_status': <PlantMeasurementStatus.NO_DATA: 0>,
      'user_picture_path': '<url>',
      'user_thumb_path': '<url>',
    }),
  })
# ---
# name: test_get_plant_list[asyncio]
  dict({
  })
# ---
